### PR TITLE
change in validation and set_python_target function

### PR DESCRIPTION
**Pull request by gourav-kandoria**

### CODE REVIEW

File: src/nxt_conf_validation.c [hunk 1]
```diff
@@ -839,6 +839,12 @@ static nxt_conf_vldt_object_t  nxt_conf_vldt_python_members[] = {
         .validator  = nxt_conf_vldt_targets_exclusive,
         .u.string   = "callable",
     }, {
+        .name       = nxt_string("factory"),
+        .type       = NXT_CONF_VLDT_STRING,
+        // need to add validator here which specify it needs to be mutually exclusive
+        // with callable and target option. not sure at present how
+    },
+    {
```
Comment (ac000) [inline src/nxt_conf_validation.c:847]: These should be on the same line...

File: src/nxt_conf_validation.c [hunk 2]
```diff
@@ -866,12 +872,14 @@ static nxt_conf_vldt_object_t  nxt_conf_vldt_python_target_members[] = {
         .name       = nxt_string("prefix"),
         .type       = NXT_CONF_VLDT_STRING,
         .validator  = nxt_conf_vldt_python_prefix,
+    }, {
+        .name       = nxt_string("factory"),
+        .type       = NXT_CONF_VLDT_STRING,
+        // need to add validator here which specify it needs to be mutually exclusive
+        // with callable option. not sure at present how
     },
```
Comment (ac000) [inline src/nxt_conf_validation.c:880]: You may as well put this above prefix so the order matches as above...

File: src/python/nxt_python.c [hunk 3]
```diff
@@ -428,27 +429,42 @@ nxt_python_set_target(nxt_task_t *task, nxt_python_target_t *target,
...
+        factory_name[str.length] = '\0';
+        nxt_alert(task, "factory name %s", factory_name);
+        obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
+        obj = PyObject_CallObject(obj, NULL);
     }
-
-    obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
     if (nxt_slow_path(obj == NULL)) {
         nxt_alert(task, "Python failed to get \"%s\" from module \"%s\"",
                   callable, module_name);
```
Comment (ac000) [inline src/python/nxt_python.c:448]: What is the value of callable now?

File: src/python/nxt_python.c [hunk 4]
```diff
@@ -428,27 +429,42 @@ nxt_python_set_target(nxt_task_t *task, nxt_python_target_t *target,
...
+        } else {
+            nxt_conf_get_string(value, &str);
+
+            callable = nxt_alloca(str.length + 1);
+            nxt_memcpy(callable, str.start, str.length);
+            callable[str.length] = '\0';
+        }
+        obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
+    }
+
```
Comment (ac000) [inline src/python/nxt_python.c:467]: I have a feeling all this logic could be cleaned up and possibly removing the need for the factory_found boolean...

**Review by ac000 — COMMENTED** [review]
Thanks for the patch!

However the commit message is missing (a few things)

1) A description of what this patch does
2) Why it's needed
3) Examples of valid config using this new feature

I'm guessing it's in relation to <https://github.com/nginx/unit/issues/1106>

It could do with a more accurate subject, `python: Add 'factory' support` or some such...

However this doesn't seem to implement @callahad's [suggestion](https://github.com/nginx/unit/issues/1106#issuecomment-2178415423). Some text describing the reasoning wouldn't go amiss...

**Comment by callahad** [comment]
It does match my [earlier suggestion](https://github.com/nginx/unit/issues/1106#issuecomment-2177162762) of

> If you have a patch, would you mind opening a pull request? Even if it's unpolished, having something concrete is a great starting point.

😆

Great way to get the ball rolling; thank you @gourav-kandoria! Are you up for refining this to match the later proposal (a separate `factory` boolean in the config?)

**Comment by gourav-kandoria** [comment]
@callahad @ac000  oh my bad really sorry . @callahad I had these changes earlier in my local. raised pr with those changes.
here is new pr as you suggested in issue comments
https://github.com/nginx/unit/pull/1336/files

**Comment by callahad** [comment]
Great, thank you! Closing in favor of #1336

**Pull request closed without merging**